libraries/AP_NavEKF3/AP_NavEKF3.cpp

```diff
@@ -815,34 +821,53 @@ void NavEKF3::UpdateFilter(void)
+        updateCoreErrors(primaryErrorScore);
 
+        bool betterCore = false;
+        bool altCoreAvailable = false;
+        float maxCoreError = relativeCoreError[primary]; // looking for cores that have error lower than the current primary
+        uint8_t newPrimaryIndex = primary;
+
+        // loop through all available cores to find if an alternative core is available
+        for (uint8_t coreIndex=0; coreIndex<num_cores; coreIndex++) {
             if (coreIndex != primary) {
-                // an alternative core is available for selection only if healthy and if states have been updated on this time step
-                bool altCoreAvailable = core[coreIndex].healthy() && statePredictEnabled[coreIndex];
-
-                // If the primary core is unhealthy and another core is available, then switch now
-                // If the primary core is still healthy,then switching is optional and will only be done if
-                // a core with a significantly lower error score can be found
-                float altErrorScore = core[coreIndex].errorScore();
-                if (altCoreAvailable && (!core[newPrimaryIndex].healthy() || altErrorScore < lowestErrorScore)) {
+                float altCoreError = relativeCoreError[coreIndex];
+
+                // an alternative core is available for selection based on 2 conditions -
+                // 1. healthy and states have been updated on this time step
+                // 2. has relative error less than primary core error
+                altCoreAvailable = core[coreIndex].healthy() && statePredictEnabled[coreIndex] && altCoreError < maxCoreError;
+
+                if(altCoreAvailable) {
+                    // if this core has a significantly lower relative error to the active primary, we consider it as a 
+                    // better core and would like to switch to it even if the current primary is healthy
+                    betterCore = (0.5*CORE_ERR_LIM + altCoreError) < relativeCoreError[primary]; // consider a better core if its relative error is lower by half the max possible error
+                    maxCoreError = altCoreError;
                     newPrimaryIndex = coreIndex;
-                    lowestErrorScore = altErrorScore;
-                }
+                } 
             }
         }
-        // update the yaw and position reset data to capture changes due to the lane switch
-        if (newPrimaryIndex != primary) {
+        altCoreAvailable = newPrimaryIndex != primary;
+
+        // Switch cores if another core is available and the active primary core meets one of the following conditions - 
+        // 1. has a bad error score
+        // 2. is unhealthy
+        // 3. is healthy, but a better core is available
+        // also update the yaw and position reset data to capture changes due to the lane switch
+        if (altCoreAvailable && (primaryErrorScore > 1.0f || !core[primary].healthy() || betterCore)) {
```

Comment:
we've lost the send_text() for the lane switch which we have in master

Reply:
master doesn't seem to have a send_text() in UpdateFilter(), its in checkLaneSwitch() here - https://github.com/ArduPilot/ardupilot/blob/ca0ae57b56f472ef656ae82eb0a8afe42217c2de/libraries/AP_NavEKF3/AP_NavEKF3.cpp#L899
I'll copy it over. Does this mean that the Lane Switch messages we were seeing before were not because of switches that happened in UpdateFilter() ? I think we might to rework checkLaneSwitch() as well for the new relative error.